src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/BaseTest.cs

```diff
@@ -151,25 +155,115 @@ public void AfterAllTests ()
+		{
+			RunAdbCommand ($"shell {command} {string.Join (" ", args)}");
+		}
+
+		protected static string ClearAdbLogcat ()
+		{
+			return RunAdbCommand ("logcat -c");
+		}
+
+		protected static string GetAdbLogcat (int timeout = 1)
+		{
+			return RunAdbCommand ("logcat", timeout: timeout);
+		}
+
+		protected static void AdbStartActivity (string activity)
+		{
+			RunAdbCommand ($"shell am start -S -n \"{activity}\"");
+		}
+
+		protected static bool MonitorAdbLogcat (Func<string, bool> action, int timeout = 10)
+		{
+			string ext = Environment.OSVersion.Platform != PlatformID.Unix ? ".exe" : "";
+			string adb = Path.Combine (AndroidSdkPath, "platform-tools", "adb" + ext);
+			var proc = Process.Start (new ProcessStartInfo (adb, "logcat") { RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false });
+			var sw = new Stopwatch ();
+			try {
+				TimeSpan time = TimeSpan.FromSeconds (timeout);
+				while (time.TotalMilliseconds > 0) {
+					sw.Start ();
+					if (action (proc.StandardOutput.ReadLine ()))
+						return true;
+					time = time.Subtract (TimeSpan.FromMilliseconds (sw.ElapsedMilliseconds));
+					sw.Reset ();
+				}
+			} finally {
+				sw.Stop ();
+				proc.Kill ();
+				proc.WaitForExit ();
+			}
+			return false;
+		}
+
+		protected static bool WaitForActivityToStart (string activityNamespace, string activityName, out string output, int timeout = 60)
+		{
+			var sb = new StringBuilder ();
+			bool result = MonitorAdbLogcat ((line) => {
+				sb.AppendLine (line);
+				return line.Contains ("ActivityManager: Displayed") && line.Contains (activityNamespace) && line.Contains (activityName);
+			}, timeout: timeout);
+			output = sb.ToString ();
+			return result;
+		}
+
+		protected static (int x, int y, int w, int h) GetControlBounds (string packageName, string uiElement, string text)
+		{
+			var regex = new Regex (@"[(0-9)]\d*", RegexOptions.Compiled);
+			var result = (x: 0, y: 0, w: 0, h: 0);
+			var ui = RunAdbCommand ("exec-out uiautomator dump /dev/tty");
+			while (ui.Contains ("ERROR:")) {
+				ui = RunAdbCommand ("exec-out uiautomator dump /dev/tty");
+				WaitFor (1);
+			}
+			ui = ui.Replace ("UI hierchary dumped to: /dev/tty", string.Empty).Trim ();
+			try {
+				var uiDoc = XDocument.Parse (ui);
+				var node = uiDoc.XPathSelectElement ($"//node[contains(@resource-id,'{uiElement}')]");
+				if (node == null)
+					node = uiDoc.XPathSelectElement ($"//node[contains(@text,'{text}')]");
+				if (node == null)
+					return result;
+				var bounds = node.Attribute ("bounds");
+				var matches = regex.Matches (bounds.Value);
+				int.TryParse (matches [0].Value, out int x);
+				int.TryParse (matches [1].Value, out int y);
+				int.TryParse (matches [2].Value, out int w);
+				int.TryParse (matches [3].Value, out int h);
+				return (x: x, y: y, w: w, h: h);
+			} catch (Exception ex) {
+				// Ignore any error and return and empty
+				throw new Exception ($"{ex.Message} {ui}");
```

Comment:
Why throw a new exception instead of letting the original through?

This shouldn't discard the original exception, *at minimum* it should provide `ex` as a nested exception.  This should also throw a "better" type than `Exception`, e.g. `InvalidOperationException`.